### DISCUSSION TITLE
Links in markdown previews now open in new tabs and result in 404

### DISCUSSION THREAD

**Discussion by ChaDonSom** [coder/code-server#3591] [Q&A] [answered]
Hi everyone, I'm having some sort of problem, but I don't know enough about my environment to be sure this is a real issue, so I thought I'd just ask here first.

I'm not sure when this started happening, but it was pretty recently after updating to 3.10.2.

Whenever I click a link in a markdown file written like `[file](./file.md)` from the preview, it comes up with a url like `http://penguin.linux.test:8083/webview/vscode-resources/file///home/chase/project/file.md` and the url results in a `404` error.

For some reason, it also is treating that url as an external url? That is, it asked first whether I wanted to trust that 'external' domain (that's the correct domain for where I'm running code-server).

Interestingly, if I click the 'Follow link' that shows up in the _editor_ mode of a markdown file, it works.

I haven't been able to google anything that quite fits this scenario, except possibly [this closed issue from Sep. 2019](https://github.com/cdr/code-server/issues/996). I've been able to run VS Code 1.56 natively (I'm running on chromebook and code-server simply has better handling) and saw that markdown-preview links work normally there (although it doesn't entirely make sense to me why they wouldn't).

Any ideas or direction toward possible relevant code I could check out or something would be much appreciated! At the same time, this is understandably not a huge priority, either.

Thank you and I hope your day goes well!

**Comment by jsjoeio** [comment, marked as answer]
Hmm...are you using any Markdown extensions by chance? 

Can this be reproduced in a code-server environment with no extensions installed? 

It seems to be working for me locally here (running 3.10.2 on macOS + Chrome):

## Screenshot

https://user-images.githubusercontent.com/3806031/121565595-70a2d700-c9d1-11eb-81fe-4f858e11ef98.mov

I checked inside VSCode and it opens the file inside the editor. I guess we default to opening the raw file in the browser (but it could be nice to open directly in code-server. Not sure if we can do that yet though).

@code-asher might have some insight too

**Reply by code-asher** [reply]
Ah yeah when I investigated this in the OP's linked issue it seemed we had parity but things might have changed. There is some code in that issue to fix the issue that might still be valid although this seems like a good candidate for a fix upstream.

**Reply by ChaDonSom** [reply]
Yes, I actually have the Github markdown styling extension by bierner: 
![image](https://user-images.githubusercontent.com/24794769/121574668-8249a600-c9f4-11eb-9993-294d3e7655a6.png)

(I added the other two after discovering this behavior)

When you mention that it defaults to viewing the raw file in the browser, I wonder if that may be what's going on. In the past, the links would open in the markdown preview pane (somehow or another it was working 🤷 😄 ).

I tried once to 'Disable All Installed Extensions': 
![image](https://user-images.githubusercontent.com/24794769/121574937-c50b7e00-c9f4-11eb-8059-556b9808ff73.png)
but I didn't notice anything change, and it may not have even actually disabled them. All the extensions had the 'Reload Required' button showing regardless of reloading.

**Reply by ChaDonSom** [reply]
I'll try it with no extensions installed when I get a chance 👍

**Reply by jsjoeio** [reply]
Ah gotcha. I was going to the Command Palette and typing "Preview" then hitting "Markdown Preview" to open it on the side. 

> When you mention that it defaults to viewing the raw file in the browser, I wonder if that may be what's going on

Yeah, I think you're right. That seems to be it.

Well, not sure if this is a bug or not 😅

Do you have access to Codespaces by chance? If you can reproduce there, then it may be a valid bug upstream.

**Reply by ChaDonSom** [reply]
I had forgotten about Codespaces, but it sure is intriguing! Just now signed up 😄

I was able to replicate on a docker container of `codercom/code-server`:
https://user-images.githubusercontent.com/24794769/121593820-4de0e480-ca0a-11eb-9713-3f71c41b75f1.mp4

But I guess my question has been taken care of now 🙂 Now I just need to figure out how in the world it was working before. It sounds like opening it in preview-mode in the editor has never been supported 🤔🤷 I do wonder if I might be able to write an extension that does that.

At any rate, thank you for responding so quickly and with so much understanding!

**Reply by ChaDonSom** [reply]
It may be worth it to note how I think my use-case might be different than average: I have code-server installed on an Ubuntu desktop where it just runs on `0:8083`. From there, I `ssh ubuntu-desktop -N -L 0:0:0:0:8083:127.0.0.1:8083` to make Chrome OS decide it's safer than `192.168.1.7:8083` and allow me to use it without security notices.

**Reply by ChaDonSom** [reply]
Yeah, I think that might be it 😅

I just thought to go ahead and try it directly from the Ubuntu desktop and it worked just like you said it was supposed to.

It must be that ssh port linking I'm doing.

**Reply by jsjoeio** [reply]
> I just thought to go ahead and try it directly from the Ubuntu desktop and it worked just like you said it was supposed to.

Oh man! Nice sleuthing! I was wondering if that could be related. Welp, I guess this is an environment issue and not a code-server issue? Either way, it would be nice if it just worked.

**Reply by ChaDonSom** [reply]
I guess it didn't work for my locally-run code-server though, so perhaps it has more to do with Chrome OS's implementation of Linux?

The weirdest thing was just that it used to just open the previews there in the same preview pane. 🤷🏼‍♀️🤔😆

But I'd reckon that, too, may have been a result of being on Chrome OS.

**Reply by jsjoeio** [reply]
Maybe you're right? We have seen some weird issues with Chrome OS's implementation of Linux. I'm not sure what could have changed — either in VSCode/code-server or something browser-related 🤔 

But at least you are unblocked (I hope 🤞)

**Reply by ChaDonSom** [reply]
I've gotten my answer, anyway 🙂

I wish it could open the preview in the same pane, but I see that it was just a happy accident due to my environment that it ever did that 😅

I definitely hope it can be done someday in the future! But also I might look into how to do it with an extension or something. But for now, I'll just click the links from the editor pane.

Thanks again for your help!
